pkg/k8s/resource/resource.go

```diff
@@ -483,21 +491,27 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
 			switch entry := entry.(type) {
 			case syncEntry:
 				event.Kind = Sync
-			case deleteEntry:
-				event.Kind = Delete
-				event.Key = entry.key
-				event.Object = entry.obj.(T)
-			case upsertEntry:
+			case upsertOrDeleteEntry:
 				obj, exists, err := store.GetByKey(entry.key)
-				// If the item didn't exist, then it's been deleted and a delete event will
-				// follow soon.
-				if err != nil || !exists {
-					event.Done(nil)
-					continue loop
+				if !exists || err != nil {
```

Comment:
I think this is still racy.

Imagine the sequence `create (C) delete (D) create (C) delete (D)` for the same resource coming from k8s. The end state should be either four, two or no events emitted (CDCD, CD or nothing), but never starting with a D (deletion of not created resource), or ending with C (end state wrong, created instead of deleted). 
However, I claim that the current code can result in `DD`, i.e. double deletion of a resource we have not emitted a create for. The key queue coalesces `upsertOrDeleteEntry{"name"} as long as we don't call `Get`. Hence, I can choose to let it coalesce the `CD` to just one, with which we enter the loop. Before we get to `store.GetByKey`, however, the third and fourth action `CD` have hit the store, which now does not contains the resource, and `deletedObjs` does. Hence, we emit just `D`, deleting a resource we didn't create.
Since the object in the `deletedObj` doesn't match, in the call to event.Done we do not remove it. The next call to `kq.Get()` thus relivers the `upsertOrDeleteEntry{"name"}` coalesced from the third and fourth action, and we emit another `D` event, since the store doesn't carry the resource, and the deletedObjs still do.

Reply:
Agreed on this. I'll look into tracking the emitted creations so deletions will always pair up with the create.

Reply:
Fixed. Creations are now tracked and Resource[T] will only emit a delete event for objects that the subscriber has observed, e.g. if `create X`, `delete X` occurs in quick succession and the subscriber didn't see the `create X` it won't see the `delete X`.